src/Orleans.Core/Providers/LegacyProviderConfigurator.cs

```diff
@@ -0,0 +1,162 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Orleans.LogConsistency;
+using Orleans.Runtime.Configuration;
+using Orleans.Storage;
+using Orleans.Streams;
+using System.Collections.Generic;
+using System.Linq;
+using Orleans.Runtime;
+
+namespace Orleans.Providers
+{
+    public class LegacyProviderConfigurator
+    {
+        public const string InitStageName = "ProviderInitStage";
+        public const string StartStageName = "ProviderStartStage";
+        internal const string SchedulingContextName = "Fallback";
+        internal delegate Task QueueTask(Func<Task> taskFunc);
+    }
+
+    internal static class LegacyProviderConfigurator<TLifeCycle>
+        where TLifeCycle : ILifecycleObservable
+    {
+        /// <summary>
+        /// Legacy way to configure providers. Will need to move to a legacy package in the future
+        /// </summary>
+        /// <returns></returns>
+        internal static void ConfigureServices(IDictionary<string, ProviderCategoryConfiguration> providerConfigurations, IServiceCollection services, int defaultInitStage, int defaultStartStage)
```

Comment:
This is the heart of this change.  This is where we wrap provider construction and initialization into lifecycle participants for initialization and register them as named services.